[v2] Added Azure Log Analytics scaler

spoplavskiy opened the pull request:
### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO)
- [X] Tests have been added
- [X] A [PR 244](https://github.com/kedacore/keda-docs/pull/244) is opened to update the documentation on https://github.com/kedacore/keda-docs
- [x] Changelog has been updated

This Scaler related to Feature Request [1061](https://github.com/kedacore/keda/issues/1061)
It implements all features and use-cases described there.

Signed-off-by: Sergiy Poplavskyi <spopla@microsoft.com>

tomkerkhove commented:
Can you update https://github.com/kedacore/keda/blob/v2/CHANGELOG.md as well please?

spoplavskiy commented:
> Can you update https://github.com/kedacore/keda/blob/v2/CHANGELOG.md as well please?

@tomkerkhove , thanks for pointing, added in last commit

zroubalik reviewed (commented):
@spoplavskiy thanks for the contribution! I don't feel confident enough to review Azure related scalers, will keep it for the others. But I wonder whether it is doable to add e2e test for this type of scaler?

https://github.com/kedacore/keda/tree/v2/tests

spoplavskiy commented:
> @spoplavskiy thanks for the contribution! I don't feel confident enough to review Azure related scalers, will keep it for the others. But I wonder whether it is doable to add e2e test for this type of scaler?
> 
> https://github.com/kedacore/keda/tree/v2/tests

@zroubalik , yes, I can try. The thing is, that Log Analytics workspace should be created on Azure subscription, used by your build pipeline. Also, a Service Principal should have access to this workspace. I see in ".env", that there is variables, related to Service Principal. Can I introduce "TEST_LOG_ANALYTICS_WORKSPACE_ID" variable? I will commit empty value (same, as others), but it should be created and added in Build Pipeline.

zroubalik commented:
Great! I think that @ahmelsayed or @jeffhollan could help you with that

spoplavskiy commented:
> Great! I think that @ahmelsayed or @jeffhollan could help you with that

@zroubalik , I have added E2E tests from Log Analytics scaler. Also, added support for 'FromEnv' definitions. I had a conversation with @ahmelsayed, provided him all required scripts to create LA workspace and grant access to Service Principal on subscription, used for tests.

tomkerkhove commented:
> > Great! I think that @ahmelsayed or @jeffhollan could help you with that
> 
> @zroubalik , I have added E2E tests from Log Analytics scaler. Also, added support for 'FromEnv' definitions. I had a conversation with @ahmelsayed, provided him all required scripts to create LA workspace and grant access to Service Principal on subscription, used for tests.

Thanks @spoplavskiy!

zroubalik commented:
@spoplavskiy this is great, e2e tests are really important, thanks! I'll keep the final review on @ahmelsayed 

Just a minor nit, could you please rebase the PR, there's a conflict in the Changelog.

spoplavskiy commented:
> @spoplavskiy this is great, e2e tests are really important, thanks! I'll keep the final review on @ahmelsayed
> 
> Just a minor nit, could you please rebase the PR, there's a conflict in the Changelog.

@zroubalik , done. But now it looks like I break static check. For some reason "pre-commit" tool thinks, that I have changed test file fully, and fails on check. Is that OK in this case?

zroubalik commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,492 @@
+		return tokenInfo, nil
+	}
+
+	logAnalyticsLog.Info(fmt.Sprintf("Reuse existing access token..."))
```
No need to call fmt.Sprintf() inside.

spoplavskiy replied:
Thanks for pointing. I have removed this entry in last commit, as it is not so valuable in this case.

zroubalik commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,492 @@
+
+	//Handle expired token
+	if statusCode == 403 || strings.Contains(string(body), "TokenExpired") {
+		logAnalyticsLog.Info(fmt.Sprintf("Token expired, refreshing token..."))
```
No need to call `fmt.Sprintf()`

spoplavskiy replied:
Fixed in last commit

zroubalik commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,492 @@
+		return metricsData{}, err
+	}
+
+	logAnalyticsLog.Info(fmt.Sprintf("Providing metric value: %d", metricsInfo.value))
```
We are using [Structured logging](https://www.client9.com/structured-logging-in-golang/).

Would be better to replace the call to something like: 
logAnalyticsLog.Info("Getting metric value, "metrics value", metricsInfo.value))

zroubalik replied:
If this is DEBUG log message, would be better to call it `logAnalyticsLog.V(1).Info("Getting metr...)`
So we don't flood output with this particular log

spoplavskiy replied:
Thanks for pointing. I have changed this definition in last commit.

zroubalik commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,492 @@
+	if currentTimeSec < tokenInfo.NotBefore {
+		if currentTimeSec < tokenInfo.NotBefore+10 {
+			sleepDurationSec := int(tokenInfo.NotBefore - currentTimeSec + 1)
+			logAnalyticsLog.Info(fmt.Sprintf("AAD token not ready, waiting %d seconds before continue...", sleepDurationSec))
```
Same as the others log messages

spoplavskiy replied:
Fixed in last commit

zroubalik reviewed (changes requested):
Minor nits on logging

spoplavskiy commented:
> Minor nits on logging

@zroubalik , thanks for pointing. All your suggestions has been implemented in last commit.

ahmelsayed commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,490 @@
+
+		s.cache.metricValue = receivedMetric.value
+
+		if receivedMetric.threshold > 0 {
+			s.cache.metricThreshold = receivedMetric.threshold
+		} else {
+			s.cache.metricThreshold = s.metadata.threshold
```
what does `receivedMetric.threshold` represent here? All other scalers use the threshold value defined in the metadata for updating the HPA config. I'm trying to understand if supplying the threshold value from Azure Log Analytics makes sense or not here.

spoplavskiy replied:
There is a possibility to set threshold, based on query result. If query returns row with 2 cells, 2d cell will be interpreted as threshold. Query result has higher priority, than threshold, set in metadata. With this, threshold in metadata can be overwritten. I described this logic and limitations in a documentation for scaler.

ahmelsayed replied:
That makes sense. I just took a look at the docs PR. Thanks for the detailed explanation there.

ahmelsayed commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,490 @@
+		return metricsData{}, fmt.Errorf("Can't decode JSON from Log Analytics REST API result: %v", err)
+	}
+
+	if statusCode == 200 {
```
nit: can this be changed into an early return. i.e: 

```
if statusCode != 200 {
    return .....
}
```

spoplavskiy replied:
Thanks for catching. Moved it upper in last commit.

ahmelsayed commented on pkg/scalers/azure_log_analytics_scaler.go:
```diff
@@ -0,0 +1,490 @@
+			return metricsData{}, fmt.Errorf("Too many rows in query result: %d. Expected: 1", len(queryData.Tables[0].Rows))
+		}
+
+		if len(queryData.Tables[0].Rows) == 1 {
```
wouldn't this always be true based on the pre-validation above?

spoplavskiy replied:
Yes, you are right, changed it in last commit. Thanks for catching.

ahmelsayed commented:
I added the required secret to the pipeline. Just have a couple of minor comments on the PR.

spoplavskiy commented:
> I added the required secret to the pipeline. Just have a couple of minor comments on the PR.

@ahmelsayed , thanks!
